File: py/ollama_total.py
#!/usr/bin/env python3
import re
import subprocess
from decimal import Decimal


def get_ollama_total():
    """
    Calculate the total storage used by all Ollama models.
    Returns a string with the total size in the most appropriate unit.
    """
    try:
        # Run the ollama list command
        result = subprocess.run(
            ["ollama", "list"], capture_output=True, text=True, check=True
        )

        # Get the output
        output = result.stdout

        # Skip first line (header)
        lines = output.strip().split("\n")[1:]

        # Regular expression to match size values like "2.7 GB", "17 GB", "800 MB", etc.
        size_pattern = r"(\d+(?:\.\d+)?) (MB|GB|TB)"

        total_bytes = Decimal("0")

        # Process each line to extract and sum the sizes
        for line in lines:
            match = re.search(size_pattern, line)
            if match:
                size_value = Decimal(match.group(1))
                size_unit = match.group(2)

                # Convert to bytes for consistent calculation
                if size_unit == "MB":
                    total_bytes += size_value * Decimal("1000000")
                elif size_unit == "GB":
                    total_bytes += size_value * Decimal("1000000000")
                elif size_unit == "TB":
                    total_bytes += size_value * Decimal("1000000000000")

        # Convert back to the most appropriate unit
        if total_bytes >= Decimal("1000000000000"):
            return f"{(total_bytes / Decimal('1000000000000')).quantize(Decimal('0.1'))} TB"
        elif total_bytes >= Decimal("1000000000"):
            return (
                f"{(total_bytes / Decimal('1000000000')).quantize(Decimal('0.1'))} GB"
            )
        elif total_bytes >= Decimal("1000000"):
            return f"{(total_bytes / Decimal('1000000')).quantize(Decimal('0.1'))} MB"
        else:
            return f"{total_bytes} B"
    except subprocess.CalledProcessError as e:
        return f"Error running ollama list: {e}"
    except Exception as e:
        return f"Unexpected error: {e}"


if __name__ == "__main__":
    print(get_ollama_total())

File: js/ollama_total.js
#!/usr/bin/env -S deno run --allow-run

/**
 * Calculate the total storage used by all Ollama models
 * @returns {string} Total size with appropriate unit
 */
async function getOllamaTotal() {
  try {
    // Run the ollama list command
    const command = new Deno.Command("ollama", {
      args: ["list"],
      stdout: "piped",
    });

    const { stdout } = await command.output();
    const output = new TextDecoder().decode(stdout);

    // Skip the header line and process the output
    const lines = output.trim().split("\n").slice(1);

    // Regular expression to match size values like "2.7 GB", "17 GB", "800 MB", etc.
    const sizePattern = /(\d+(?:\.\d+)?) (MB|GB|TB)/;

    let totalBytes = 0;

    // Process each line
    for (const line of lines) {
      const match = line.match(sizePattern);
      if (match) {
        const sizeValue = parseFloat(match[1]);
        const sizeUnit = match[2];

        // Convert to bytes for consistent calculation
        switch (sizeUnit) {
          case "MB":
            totalBytes += sizeValue * 1000000;
            break;
          case "GB":
            totalBytes += sizeValue * 1000000000;
            break;
          case "TB":
            totalBytes += sizeValue * 1000000000000;
            break;
        }
      }
    }

    // Convert back to the most appropriate unit with one decimal place
    if (totalBytes >= 1000000000000) {
      return `${(totalBytes / 1000000000000).toFixed(1)} TB`;
    } else if (totalBytes >= 1000000000) {
      return `${(totalBytes / 1000000000).toFixed(1)} GB`;
    } else if (totalBytes >= 1000000) {
      return `${(totalBytes / 1000000).toFixed(1)} MB`;
    } else {
      return `${totalBytes} B`;
    }
  } catch (error) {
    return `Error: ${error.message}`;
  }
}

// Output the result when run directly
if (import.meta.main) {
  console.log(await getOllamaTotal());
}

// Export for use as a module
export { getOllamaTotal };
